File: src/RouterWizzard/Commands/Ubiquiti/UbiquitiClient.cs
using System;
using RouterWizzard.Commands.Ssh;

namespace RouterWizzard.Commands.Ubiquiti
{
    internal class UbiquitiClient : IUbiquitiClient
    {
        private readonly ISshClient sshClient;

        public UbiquitiClient(ISshClient sshClient)
        {
            this.sshClient = sshClient ?? throw new ArgumentNullException(nameof(sshClient));
        }

        public void BeginSession()
        {
            var result = Execute("begin");
        }

        public void EndSession()
        {
            var result = Execute("end");
        }

        public string Show(string key)
        {
            var result = Execute($"show {key}");
            return result;
        }

        public void Set(string key)
        {
            var result = Execute($"set {key}");
        }

        public void Delete(string key)
        {
            var result = Execute($"delete {key}");
        }

        public void Commit()
        {
            var result = Execute("commit");
        }

        public void Save()
        {
            var result = Execute("save");
        }

        public void ExecuteInSavedSession(Action codeToExecuteFunc)
        {
            this.BeginSession();

            codeToExecuteFunc();

            this.Commit();
            this.Save();
            this.EndSession();
        }

        private string Execute(string key)
        {
            var command = $"/opt/vyatta/sbin/vyatta-cfg-cmd-wrapper {key}";
            return this.sshClient.Execute(command);
        }
    }
}

File: src/GeoBypasser/GeoBypasser/Clients/UbiquitiClient.swift
//
//  UbiquitiClient.swift
//  RouterWizzard
//
//  Created by Vladimir Aubrecht on 07/05/2020.
//  Copyright © 2020 Vladimir Aubrecht. All rights reserved.
//

import Foundation

class UbiquitiClient {
    private let sshClient: SshClient
    
    init(sshClient: SshClient) {
        self.sshClient = sshClient
    }
    
    public func show(key: String) throws -> String {
        return try self.execute(key: "show \(key)")
    }
    
    public func delete(key: String) throws {
        _ = try self.execute(key: "delete \(key)")
    }
    
    public func set(key: String) throws {
        _ = try self.execute(key: "set \(key)")
    }
    
    public func beginSession() throws {
        _ = try self.execute(key: "begin")
    }
    
    public func endSession() throws {
        _ = try self.execute(key: "end")
    }
    
    public func commit() throws {
        _ = try self.execute(key: "commit")
    }
    
    public func save() throws {
        _ = try self.execute(key: "save")
    }
    
    private func execute(key: String) throws -> String{
        let command = "/opt/vyatta/sbin/vyatta-cfg-cmd-wrapper \(key)";
        return try self.sshClient.execute(command: command);
    }
}
